database/models.go

```diff
@@ -19,50 +19,121 @@ import (
 	"encoding/json"
 	"time"
 
+	"github.com/deckarep/golang-set"
+
 	"github.com/coreos/clair/pkg/pagination"
 )
 
-// Processors are extentions to scan a layer's content.
-type Processors struct {
-	Listers   []string
-	Detectors []string
-}
-
 // Ancestry is a manifest that keeps all layers in an image in order.
 type Ancestry struct {
 	Name string
 	// ProcessedBy contains the processors that are used when computing the
 	// content of this ancestry.
-	ProcessedBy Processors
+	ProcessedBy []Detector
 	// Layers should be ordered and i_th layer is the parent of i+1_th layer in
 	// the slice.
 	Layers []AncestryLayer
 }
 
 // AncestryLayer is a layer with all detected namespaced features.
 type AncestryLayer struct {
-	LayerMetadata
-
+	Hash string
 	// DetectedFeatures are the features introduced by this layer when it was
 	// processed.
-	DetectedFeatures []NamespacedFeature
+	DetectedFeatures []DetectedNamespacedFeature
 }
 
-// LayerMetadata contains the metadata of a layer.
-type LayerMetadata struct {
-	// Hash is content hash of the layer.
-	Hash string
-	// ProcessedBy contains the processors that processed this layer.
-	ProcessedBy Processors
+func (l *AncestryLayer) GetNamespacedFeatures() []NamespacedFeature {
+	nsf := make([]NamespacedFeature, 0, len(l.DetectedFeatures))
+	for _, f := range l.DetectedFeatures {
+		nsf = append(nsf, f.NamespacedFeature)
+	}
+
+	return nsf
+}
+
+// DetectedNamespacedFeature is a namespaced feature with the detectors used to
+// find this feature.
+type DetectedNamespacedFeature struct {
+	NamespacedFeature
+
+	// FeatureBy is the detector that detected the feature.
+	FeatureBy Detector
+	// NamespaceBy is the detector that detected the namespace.
+	NamespaceBy Detector
 }
 
-// Layer is a layer with its detected namespaces and features by
-// ProcessedBy.
+// DetectedNamespaces contains a map from detector to the detected namespace.
+// Nil detected namespace means it's not found.
+type DetectedNamespaces map[Detector]*Namespace
```

Comment:
The new map type introduces a lot more complexity. It makes the worker complicated and also the database implementation. Maybe, we don't want this map to organize since the map will at most contain around 1000 features.

Reply:
Flatten the map to be []DetectedNamespace, which contains Detector and Namespace.
This is for simplifying the data structure, we can then use pointers and a new data structure if we need to improve the performance.